GHA: Add clang 15, 16 & 17

Flamefire opened the pull request.

Flamefire commented:
@pdimov I've seen you used a container for Ubuntu 23 in https://github.com/boostorg/container_hash/commit/d9cb23d96ae22eaf843597d4aa144d6bd2d2ac35   
Any comments on this using the native image? (I'd assume using a container incurs a slight performance penalty)

pdimov commented:
The native 23.04 image probably didn't exist yet when I added gcc-13/clang-16, which is why I used a container.

Given the tendency of the native images to periodically change and break things, I wonder whether we need to just bite the bullet and use containers everywhere to ensure a stable environment. I haven't observed any significant performance penalty (although I haven't been looking very hard.)

pdimov commented:
Looks like it still doesn't exist. :-)

pdimov commented:
While we're here we may also try clang-17 using ubuntu:23.10.

Flamefire commented:
Yep, looks like the image name is accepted but there are no runners. Moved into a container. I needed to use #220 for that though.

Also added Clang 17 as suggested.
